### PR TITLE
[Feature] Add rating to programming exercises

### PR DESCRIPTION
**Pull request by fde312**
### Checklist
- [x] I tested *all* changes and *all* related features locally and on test server1
- [x] Client: I followed the [coding and design guidelines](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/client.html).
- [x] Client: I added multiple screenshots/screencasts of my UI changes

### Motivation and Context

As the ratings were missing for programming exercises, we add them in this PR.

### Description

- Display RatingsComponent under the ComplaintInteractionComponent (as in the other exercise types)
- Add correct router link for programming exercises

### Steps for Testing

1. Log in to Artemis
2. Navigate to Course Administration
3. Create a programming exercise with manual assessment, participate and assess  or use already assessed exercise
4. Navigate to the exercise details view and scroll down (see screenshot below)
5. Add your rating
6. Login as Instructor
7. Go to `Instructor Course Dashboard` and click on Ratings
8. The exercise you rated before should be now displayed here.
9. Click on `Show the submission` and you should be forward to the code editor (make sure it is the correct submission)

image

### Test Coverage

Nothing changed

### Screenshots

image

image

### CODE REVIEW

**Review by airelawaleria — APPROVED** [review]
Tested on ts1, everything works

File: src/main/webapp/app/exercises/shared/rating/rating.component.ts [hunk 1]
```diff
@@ -46,6 +44,8 @@ export class RatingComponent implements OnInit {
 
         // update feedback locally
         this.rating.rating = $event.newValue;
+        // Delete participation to avoid circular dependency
+        this.rating.result!.participation = undefined;
```
Comment (balazs-czopf) [inline src/main/webapp/app/exercises/shared/rating/rating.component.ts:48]: I am personally not a fan of the exclamation mark as it says "assume the property/variable **IS** defined", but what if not? There must be a reason why the result can be undefined, so I´d rather go with something like this here:
```suggestion
        if (this.rating.result) {
                this.rating.result.participation = undefined;
        }
```

File: src/main/webapp/app/exercises/shared/rating/rating.component.ts [hunk 2]
```diff
@@ -23,8 +23,6 @@ export class RatingComponent implements OnInit {
             return;
         }
 
-        // delete participation to prevent circular dependency
-        this.result.participation = undefined;
         this.ratingService.getRating(this.result.id!).subscribe((rating) => {
```
Comment (balazs-czopf) [inline src/main/webapp/app/exercises/shared/rating/rating.component.ts:26]: Same goes for the exclamation mark here as below
```suggestion
        if (this.result.id !== undefined) {
                this.ratingService.getRating(this.result.id).subscribe((rating) => {
```

**Review by balazs-czopf — CHANGES_REQUESTED** [review]
Code looks good, only added two minor comments on using exclamation marks.
If there is a specific reason why we use them which I don´t know, feel free to ignore my comments. 😄

**Review by schultek — APPROVED** [review]
Works as expected. 👍 Nice addition.

**Review by JohannesWeiss — COMMENTED** [review]
Basically works as intended, but the ratings (and in the second attempt the name) were not displayed in the ratings overview. Tested on ts1 with Safari 14.0.
image
The link to the tested repos works correctly.

**Review by simonzilker — COMMENTED** [review]
I am experiencing the same issues as @JohannesWeiss on ts3. Using Chrome Version 86.0.4240.198
image

File: src/main/webapp/app/exercises/shared/rating/rating.component.ts (hunk 1 above)
Comment (fde312) [reply src/main/webapp/app/exercises/shared/rating/rating.component.ts:48]: Done in `a92ca57`

File: src/main/webapp/app/exercises/shared/rating/rating.component.ts (hunk 2 above)
Comment (fde312) [reply src/main/webapp/app/exercises/shared/rating/rating.component.ts:26]: Done in `a92ca57`

**Review by JohannesWeiss — APPROVED** [review]
Tested with @simonzilker and now works as intended.

**Review by simonzilker — APPROVED** [review]
tested with @JohannesWeiss , no bugs anymore! Works as intended

**Review by airelawaleria — APPROVED** [review]
Everything works :)

**Review by EneaGore — APPROVED** [review]
Everything seems to work perfectly.

**Review by balazs-czopf — APPROVED** [review]
Code looks good to me now. Nice work. 🚀

**Review by krusche — APPROVED** [review]

**Pull request merged**